VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol

```diff
@@ -0,0 +1,545 @@
+            
+            if (_datacenterInfo[datacenterId].nodeCount != 0) {
+                revert HasDependingNodes();
+            }
+        }
+
+        Datacenter storage datacenter = _dataCenterMap[datacenterId];
+        
+        _countryDataCenterIds[datacenter.countryCode].remove(datacenterId);
+        _regionDataCenterIds[datacenter.regionCode].remove(datacenterId);
+        delete _dataCenterNameToID[datacenter.name];
+        delete _dataCenterMap[datacenterId];
+        delete _datacenterInfo[datacenterId];
+
+        emit RemoveDataCenter(datacenterId);
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getDatacenters(uint[] calldata ids) external view override returns(Datacenter[] memory) {
+        uint count = ids.length;
+        Datacenter[] memory list = new Datacenter[](count);
+
+        for (uint i; i < count; ++i) {
+            if (!_datacenterInfo[ids[i]].isActive) {
+                revert InvalidDatacenterId(ids[i]);
+            }
+
+            list[i] = _dataCenterMap[ids[i]];
+        }
+
+        return list;
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getDataCentersByCountry(string calldata countryCode) external view override returns(Datacenter[] memory) {
+        validateCountryCode(countryCode);
+        
+        uint count = _countryDataCenterIds[countryCode].length();
+        Datacenter[] memory list = new Datacenter[](count);
+
+        EnumerableSetUpgradeable.UintSet storage countryIds = _countryDataCenterIds[countryCode];
+
+        for (uint i; i < count; ++i) {
+            list[i] = _dataCenterMap[countryIds.at(i)];
+        }
+
+        return list;
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function getDataCentersByRegion(string calldata regionCode) external view override returns(Datacenter[] memory) {
+        validateRegionCode(regionCode);
+        
+        uint count = _regionDataCenterIds[regionCode].length();
+        Datacenter[] memory list = new Datacenter[](count);
+
+        EnumerableSetUpgradeable.UintSet storage regionIds = _regionDataCenterIds[regionCode];
+
+        for (uint i; i < count; ++i) {
+            list[i] = _dataCenterMap[regionIds.at(i)];
+        }
+
+        return list;
+    }
+
+    /**
+     * @notice Check the `authSignature` parameter of `addNode()` function
+     * @param didAddress DID address that is associated with the storage node
+     * @param authSignature Signature signed by a trusted signer
+     */
+    function verifyAuthSignature(address didAddress, bytes calldata authSignature) internal view {
+        EnumerableSetUpgradeable.AddressSet storage signers = _trustedSigners;
+
+        if (signers.length() == 0) {
+            revert NoSigners();
+        }
+
+        if (authSignature.length == 0) {
+            revert InvalidSignature();
+        }
+
+        bytes memory rawMsg = abi.encodePacked(didAddress);
```

Comment:
Do we need a nonce to prevent re-entry attacks?

Reply:
No, need in this AuthSignature verification.
`nonce` check is done in the `verifyRequest()` function.